usl_pipeline/cloud_functions/main.py

```diff
@@ -34,44 +35,93 @@ class FeatureMetadata:
     elevation_max: Optional[float] = None
 
 
+def _retry_and_report_errors(
+    func: Callable[[functions_framework.CloudEvent], None]
+) -> Callable[[functions_framework.CloudEvent], None]:
+    """Decorator which adds error handling and retries to a cloud function."""
+
+    @functools.wraps(func)
+    def wrapper(cloud_event: functions_framework.CloudEvent) -> None:
+        logging.basicConfig(level=logging.WARN)
+
+        # To handle retries, check the time created and return if the event is too old.
+        # GCP retries until the function returns successfully, so the pattern is to
+        # report exceptions to have the function retried and return normally to stop
+        # retries. https://cloud.google.com/functions/docs/bestpractices/retries
+        event_time = datetime.datetime.fromisoformat(cloud_event.data["timeCreated"])
+        event_age = (
+            datetime.datetime.now(datetime.timezone.utc) - event_time
+        ).total_seconds()
+        if event_age > _MAX_RETRY_SECONDS:
+            logging.error(
+                "Dropped event id %s source %s after %s seconds",
+                cloud_event["id"],
+                cloud_event["source"],
+                _MAX_RETRY_SECONDS,
+            )
+            return
+
+        # Catch exceptions and report them with GCP error reporting.
+        try:
+            func(cloud_event)
+        except Exception as exc:  # noqa
+            # Log the exception to display in Cloud Function logs.
+            logging.exception(exc)
+            # Report the error in GCP Error Reporter.
+            error_reporting.Client().report_exception()
+            # Write the error to the metastore.
+            _write_chunk_metastore_error(cloud_event.data["name"], str(exc))
+
+    return wrapper
+
+
+@functions_framework.cloud_event
+@_retry_and_report_errors
+def write_study_area_metadata(cloud_event: functions_framework.CloudEvent) -> None:
```

Comment:
Triggering CityCat will be a manual process. We don't want to do any automated CityCAT runs. Folks will be running the script Anna showed us manually.

We don't know if a study area being uploaded is just for predictions, so we don't know if we should run CityCAT on it or not. We also don't want to run it automatically to avoid unneeded simulation costs.

Reply:
I see. Sounds good.